Firestore/core/src/firebase/firestore/local/memory_lru_reference_delegate.h

```diff
@@ -0,0 +1,119 @@
+ * limitations under the License.
+ */
+
+#ifndef FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_MEMORY_LRU_REFERENCE_DELEGATE_H_
+#define FIRESTORE_CORE_SRC_FIREBASE_FIRESTORE_LOCAL_MEMORY_LRU_REFERENCE_DELEGATE_H_
+
+#include <memory>
+#include <unordered_map>
+#include <utility>
+
+#include "Firestore/core/src/firebase/firestore/local/lru_garbage_collector.h"
+#include "Firestore/core/src/firebase/firestore/local/reference_delegate.h"
+
+namespace firebase {
+namespace firestore {
+namespace local {
+
+class ListenSequence;
+class MemoryPersistence;
+class Sizer;
+
+/**
+ * Provides the LRU GC implementation for memory persistence.
+ */
+class MemoryLruReferenceDelegate : public LruDelegate {
+ public:
+  MemoryLruReferenceDelegate(MemoryPersistence* persistence,
+                             LruParams lruParams,
+                             std::unique_ptr<Sizer> sizer);
+
+  bool IsPinnedAtSequenceNumber(model::ListenSequenceNumber upperBound,
+                                const model::DocumentKey& key);
+
+  // MARK: ReferenceDelegate overrides
+
+  model::ListenSequenceNumber current_sequence_number() override;
+
+  void AddInMemoryPins(ReferenceSet* set) override;
+
+  void AddReference(const model::DocumentKey& key) override;
+
+  void RemoveReference(const model::DocumentKey& key) override;
+
+  void RemoveMutationReference(const model::DocumentKey& key) override;
+
+  void RemoveTarget(const QueryData& queryData) override;
+
+  void UpdateLimboDocument(const model::DocumentKey& key) override;
+
+  void OnTransactionStarted(absl::string_view label) override;
+
+  void OnTransactionCommitted() override;
+
+  // MARK: LruDelegate overrides
+
+  LruGarbageCollector* garbage_collector() override;
+
+  int64_t CalculateByteSize() override;
+
+  size_t GetSequenceNumberCount() override;
+
+  void EnumerateTargets(const TargetCallback& callback) override;
+
+  void EnumerateOrphanedDocuments(
+      const OrphanedDocumentCallback& callback) override;
+
+  int RemoveOrphanedDocuments(model::ListenSequenceNumber upper_bound) override;
+
+  int RemoveTargets(model::ListenSequenceNumber sequence_number,
+                    const LiveQueryMap& live_queries) override;
+
+ private:
+  bool MutationQueuesContainKey(const model::DocumentKey& key);
+
+  // This instance is owned by MemoryPersistence.
+  MemoryPersistence* persistence_ = nullptr;
+
+  std::unique_ptr<Sizer> sizer_;
+
+  // Tracks sequence numbers of when documents are used. Equivalent to sentinel
+  // rows in the leveldb implementation.
+  std::unordered_map<model::DocumentKey,
+                     model::ListenSequenceNumber,
+                     model::DocumentKeyHash>
+      sequence_numbers_;
+
+  // This ReferenceSet is owned by FSTLocalStore.
+  ReferenceSet* additional_references_ = nullptr;
+
+  std::unique_ptr<LruGarbageCollector> gc_;
+
+  // This needs to be a pointer because initialization is delayed until after
```

Comment:
Any particular reason not to use `DelayedConstructor`? Has it outlived its usefulness?

Reply:
Thanks for explaining. I saw `DelayedConstructor` as a vocabulary type that conveys the intent better than `optional` while having similar semantics. I'm not opposed to looking at it as a temporary compatibility shim.